src/boundary_condition.cpp

```diff
@@ -16,7 +17,19 @@ namespace openmc {
 
 void VacuumBC::handle_particle(Particle& p, const Surface& surf) const
 {
-  p.cross_vacuum_bc(surf);
+  // Random ray and Monte Carlo need different treatments at vacuum BCs
```

Comment:
I don't see why specular reflection is necessary here. Why not, for example, white reflection? It might end up sampling the angular space better that way.

Also, I'm not super aesthetically pleased about having one object be responsible for two entirely different tasks aside from being a physical representation of a vacuum BC. But I guess this is the simplest way to fit it into the current source code?

Reply:
> I don't see why specular reflection is necessary here. Why not, for example, white reflection? It might end up sampling the angular space better that way.

This is a cool idea, I had also considered this at some point, but I think a white condition might end up biasing the angular spectrum of the integrating rays. For instance, let's imagine a 300 cm pincell with full vacuum boundary conditions on all sides. What we want is a uniform ray density in space and angle. In this case, if white is used, rays traveling in a shallow azimuthal direction will immediately hit a vacuum with a cm and have their angles resampled. After a few tens of cm of active length, all integrating rays will eventually sample a very steep (highly polar) angle and then travel a long distance in that direction without colliding and resampling. Thus, even though all rays start with a random distribution of angles, after a few cm of active length the angular integrating spectrum is basically all steep polar angles, resulting in bias.


> Also, I'm not super aesthetically pleased about having one object be responsible for two entirely different tasks aside from being a physical representation of a vacuum BC. But I guess this is the simplest way to fit it into the current source code?

A fair point -- we could in theory have it generate a different BC type when launched in random ray mode (e.g., `RandomRayVacuumBC` vs `MonteCarloVacuumBC`) or something, but I thought it was slightly simpler to have the function just dispatch here based on the mode/particle type. Going that route would involve a check in the input stage as well as the two different object definitions, whereas the present method just changes a few lines in one function. Happy to go with `RandomRayVacuumBC` vs `MonteCarloVacuumBC`(or something else) though if you think it would be easier to understand!

Reply:
I do have a slight bit of anxiety about adding more and more branching code to OpenMC as we add features, since this can detrimentally impact even CPU performance. Plenty of people advocate branch-free programming on CPUs.

Hmmm, what I really wish is that OpenMC was a little bit more like a library, where the main continuous energy and multigroup transport modes are just two separate apps built using our library components. Of course random ray or MOC would be its own app built on it all. Given the structuring into classes, it's already so close to being like that!

But with the realities of the situation in mind and thinking of this as an MVP, it definitely makes sense to stick with the approach you've implemented here.